Add stereoscopic rendering through multiview

BastiaanOlij opened the pull request:
https://github.com/godotengine/godot-proposals/issues/2648

The idea here is too re-introduce stereoscopic rendering into Godot 4 but take a different approach to Godot 3. Instead of rendering one eye after the other we're going to provide multiple cameras to the render server and have it make the right decision. The initial implementation will be using multiview which means that all render buffer textures get an extra layer and we'll end up rendering left and right eye images in parallel with the Vulkan driver performing the duplication. This means we get stereoscopic rendering with a minimum of state changes and a minimum of changes to the render engine itself.

To support this there are a number of areas that are changing. 

While we are focusing on implementing stereoscopic rendering wherever possible we're building things in a way that additional views can be supported going forward. 

**Adding Multiview to render device**
While detection of Multiview capabilities was already merged this PR introduces further changes that enable the Vulkan implementation of Multiview. These include:
- proper initialization of the Vulkan device so Multiview features are enabled if available.
- creating texture arrays instead of textures for our render buffers with a layer for each view
- creation of a Multiview render buffer instead of a normal render buffer if more then one view is requested

**Changes to XRInterface**
`XRInterface` is changing to accommodate this new approach.  
- `is_stereo` is replaced with `get_view_count` which return the number of views we need to support
- `get_camera_transform` has been added to get the center position and is used to place our node and determine our reference frame when recentering our player. This replaces calling `get_transform_for_eye(EYE_MONO, ...`
- `get_transform_for_view` replaces `get_transform_for_eye` and now takes a view index 
- `get_projection_for_view` replaces `get_projection_for_eye` and now takes a view index, note that this removes the ability to call this with `EYE_MONO` but as that never worked for any stereoscopic implementation and for mono applications it returns the correct projection matrix it isn't a great loss
- `commit_eyes` replaces `commit_for_eye` and is now called after rendering of both eyes is finished with a render buffer RID that contains our render result. This functions can optionally return a blit list that tells the render engine whether output to screen is required (many VR runtimes will output to the headset only).   
- `get_external_texture_for_eye` for now is deprecated. A new implementation of this will need to be introduced.

Note, further changes are delegated to #49301

**Changes to viewports**
The `use_xr` flag (was `arvr`) has been re-enabled and will trigger the use of the primary XR interface for this viewport.
- if `use_xr` is true `get_view_count` is used to ensure the render buffer related to the viewport will contain the correct number of layers
- if `use_xr` is true `get_transform_for_view` and `get_projection_for_view` are called for each view to create the camera data send to the renderer

> 2D rendering if more then 1 view is requested is defunct. There are a few strategies but most are questionable especially if we're not looking at run of the mill stereoscopic rendering. There is an argument for disabling 2D all together when more then 1 view is in play.

**Changes to shader compilation**
Where shaders are used for stereo rendering they need to be altered to support Multiview. The changes themselves are generally minor mostly revolving around choosing a different projection matrix based on `ViewIndex`. SpirV already adds the requires support.
- if Multiview is available we declare a new define: `has_VK_KHR_multiview`
- when compiling a shader for multiview we define `MULTIVIEW` in the shader version
- uniform buffers are used to contain the additional projection matrices for each view

> Currently we define additional shader versions for stereo/multiview but it gets a little unwieldy. I'm looking into an alternative where we can create a normal and multiview version of the same shader similar to how cull variants are created

**Changes to the render engine**

*edit* note that the changes to the clustered forward renderer were moved to PR #49092, this PR will focus purely on the mobile renderer implementation.

This is where there is still a lot of work to do with only the core rendering now working properly:
- it is here where the limitation of 2 views currently exists and mostly due to limited storage for projection matrices
- culling currently uses the combined frustum that is created but both occlusion culling and light culling in the cluster implementation are broken because they do not take stereo into account
- `scene_data` now contains additional projection matrices for each view
- the main render loop in both clustered and mobile renderer now switch between normal and stereo variants of the shaders
- the sky shader has been changed to support stereo rendering
- the tonemap logic has been changed to support tonemapping all views in one go
- blitting our output to screen

> There are still many parts here that need updating, one is re-introducing external textures/render targets to render to. The other is to make an option in the mobile render to skip the copy logic and use a single render buffer (possibly the external one). This involves disabling DOF/Glow/SSR and a bunch of other post effects and moving basic tonemapping into the main shader.

We may introduce a fallback if required later on in case multiview is not supported though according to the Vulkan documentation any Vulkan 1.1 device should provide a multiview implementation.

m4gr3d commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -2469,10 +2494,9 @@ void RendererSceneCull::_frustum_cull(FrustumCullData &cull_data, FrustumCullRes
-	// Note, in stereo rendering:
-	// - p_cam_transform will be a transform in the middle of our two eyes
-	// - p_cam_projection is a wider frustrum that encompasses both eyes
+void RendererSceneCull::_render_scene(const SceneCamera *p_cameras, const uint32_t p_num_cameras, RID p_render_buffers, RID p_environment, RID p_force_camera_effects, uint32_t p_visible_layers, RID p_scenario, RID p_shadow_atlas, RID p_reflection_probe, int p_reflection_probe_pass, float p_screen_lod_threshold, bool p_using_shadows) {
```
Worth asserting that `p_num_cameras` is always `> 0`, and bailing out with some errors if for some reasons it isn't.

BastiaanOlij replied:
Will do :)

m4gr3d commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -2119,58 +2119,83 @@ bool RendererSceneCull::_light_instance_update_shadow(Instance *p_instance, cons
+		bool is_stereo = p_interface->is_stereo();
+
+		float aspect = p_viewport_size.width / (float)p_viewport_size.height;
+		cameras[0].projection = p_interface->get_projection_for_eye(is_stereo ? XRInterface::EYE_LEFT : XRInterface::EYE_MONO, aspect, camera->znear, camera->zfar);
```
I wonder if we should follow the example of the mulitview `ViewIndex` variable and have `EYE_MONO` and `EYE_LEFT` be the same value.
For one it would simplify the logic below, as you could use the matching defined constant as an index in the `SceneCamera` array.

BastiaanOlij replied:
The problem is that `EYE_MONO` and `EYE_LEFT` are not the same. `EYE_MONO` provides a center location of the head, `EYE_LEFT` is centered on the camera/eye.

So while we will have ViewIndex be 0 for Mono/Left and 1 for Right, it doesn't align with these variables. Bit of a pain I don't really have an answer for right now.

BastiaanOlij replied:
I've changed this seeing I'm looking at putting the ground work in for having more then 2 views. 

`is_stereo` is being replaced by `get_view_count`, we'll have a `get_camera_transform` to get the position and orientation of the HMD/Phone/etc and then we have `get_projection_for_view` and `get_transform_for_view` replacing resp. `get_projection_for_eye` and `get_transform_for_eye`. Both new functions take a simple `p_view` input.

BastiaanOlij commented:
Currently broken and just prototyping with loads of shortcuts but getting closer to it actually rendering in stereo. Its now creating the main render buffers as multiview buffers and compiling shaders with the main code hopefully properly rendering stuff in stereo.

Need to figure out why it currently has a mismatch in buffers causing it to fail rendering

BastiaanOlij committed:
Add stereoscopic rendering through multiview

BastiaanOlij commented:
Starting to get somewhere with this:
![image](https://user-images.githubusercontent.com/1945449/117981939-d8391880-b378-11eb-9db8-b04c40495618.png)

So it is duplicating the drawcalls nicely, but it's either rendering to both layers or it's not splitting the layers correctly.... RenderDoc is crashing with a weird error so I can't debug very properly at the moment.

I'll figure it out :)

BastiaanOlij commented:
Why RenderDoc? Why?!? :)
```
CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] RENDERDOC_MakeDebuggablePackage
[1] RENDERDOC_MakeDebuggablePackage
[2] RENDERDOC_EndProfileRegion
[3] RENDERDOC_EndProfileRegion
[4] RENDERDOC_EndProfileRegion
[5] RENDERDOC_EndProfileRegion
[6] RENDERDOC_EndProfileRegion
[7] VK_LAYER_RENDERDOC_CaptureNegotiateLoaderLayerInterfaceVersion
[8] VK_LAYER_RENDERDOC_CaptureNegotiateLoaderLayerInterfaceVersion
[9] RENDERDOC_EndProfileRegion
[10] vkCreateRenderPass (D:\Development\godot4-git\thirdparty\vulkan\loader\trampoline.c:1467)
[11] RenderingDeviceVulkan::_render_pass_create (D:\Development\godot4-git\drivers\vulkan\rendering_device_vulkan.cpp:3494)
[12] RenderingDeviceVulkan::framebuffer_format_create (D:\Development\godot4-git\drivers\vulkan\rendering_device_vulkan.cpp:3516)
[13] RenderingDeviceVulkan::framebuffer_create (
```

BastiaanOlij commented:
Working on changing things over to use texture arrays, I missed that they should be configured as such.. still not rendering properly

BastiaanOlij commented:
![image](https://user-images.githubusercontent.com/1945449/118116679-e7c66900-b42d-11eb-8d1b-f644951e4978.png)

It's now rendering correctly but only when started from RenderDoc.  Examining a frame from within RenderDoc I can see stereo is being applied correctly.

BastiaanOlij commented:
```
ERROR: VALIDATION - Message Id Number: -1480880714 | Message Id Name: VUID-VkShaderModuleCreateInfo-pCode-01091
	Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-01091 ] Object 0: handle = 0x1ba4e0882f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xa7bb8db6 | vkCreateShaderModule(): The SPIR-V Capability (MultiView) was declared, but none of the requirements were met to use it. The Vulkan spec states: If pCode declares any of the capabilities listed in the SPIR-V Environment appendix, one of the corresponding requirements must be satisfied (https://vulkan.lunarg.com/doc/view/1.2.176.1/windows/1.2-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-01091)
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_DEVICE, Handle 1899684725496
   at: VulkanContext::_debug_messenger_callback (drivers\vulkan\vulkan_context.cpp:157)

```

Currently stuck on this, when using a multiview shader something we are sending to `vkCreateGraphicsPipelines` is not correct. As far as I can see we're sending it the correct framebuffer, must be one of the other bits of the shader. 

As mentioned before, the weird thing is that it works when running using RenderDoc

BastiaanOlij commented:
With some help from people on the Vulkan Slack I was able to solve all the validation issues and this now runs properly.

Still a LOT of work to do to make everything that needs stereo support have stereo support but most of the difficult stuff is now done. Will first work on making this work on the mobile renderer too.

Would be good to get some early eyes on things because this is a BIG job that touches on a lot of parts.

reduz commented on drivers/dummy/rasterizer_dummy.h:
```diff
@@ -175,7 +175,7 @@ class RasterizerSceneDummy : public RendererSceneRender {
 	void gi_probe_set_quality(RS::GIProbeQuality) override {}
 
-	void render_scene(RID p_render_buffers, const Transform &p_cam_transform, const CameraMatrix &p_cam_projection, bool p_cam_ortogonal, const PagedArray<GeometryInstance *> &p_instances, const PagedArray<RID> &p_lights, const PagedArray<RID> &p_reflection_probes, const PagedArray<RID> &p_gi_probes, const PagedArray<RID> &p_decals, const PagedArray<RID> &p_lightmaps, RID p_environment, RID p_camera ...
+	void render_scene(RID p_render_buffers, const SceneCamera *p_cameras, const uint32_t p_num_cameras, const PagedArray<GeometryInstance *> &p_instances, const PagedArray<RID> &p_lights, const PagedArray<RID> &p_reflection_probes, const PagedArray<RID> &p_gi_probes, const PagedArray<RID> &p_decals, const PagedArray<RID> &p_lightmaps, RID p_environment, RID p_camera_effects, RID p_shadow_atlas, RID  ...
```
I have the feeling Transform should remain as-is, and only the projection matrix have some effect.

BastiaanOlij replied:
Nope, transforms are different per camera. they are about 6.5 cm apart.

BastiaanOlij replied:
This now works as discussed on Godot Chat

reduz commented on drivers/vulkan/rendering_device_vulkan.cpp:
```diff
@@ -3469,6 +3469,26 @@ VkRenderPass RenderingDeviceVulkan::_render_pass_create(const Vector<AttachmentF
+	const uint32_t correlation_mask = 0b00000011;
+	VkRenderPassMultiviewCreateInfo render_pass_multiview_create_info;
+
+	if (p_views == 2) { // only support stereo for now...
```
Then error if views > 2

reduz commented on servers/rendering/renderer_rd/effects_rd.h:
```diff
@@ -59,6 +59,12 @@
 #include "servers/rendering_server.h"
 
 class EffectsRD {
+	enum ViewMode {
```
I would rename MONO to SINGLE, MONO can be misunderstood as normalizing stereo to single channel in audio and its the same here.

BastiaanOlij replied:
Hmmm, monoscopic is incredibly common use here, we've been using the term MONO in XR for the past 3 years... I'm happy to change it but this is not audio :)

BastiaanOlij replied:
Btw @reduz , going through some older code I was reminded of this comment I put in the code around MONO V.S. STUDIO:  `my son says we should call this EYE_CYCLOPS`

In all seriousness, according to Fredia the term that seems to be used a lot in the XR space nowadays is "pancake mode".

BastiaanOlij replied:
I've removed this all together simply opting for switching between the normal shaders and multiview variants of the shaders.

reduz commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:
```diff
@@ -183,13 +183,14 @@ void RenderForwardClustered::RenderBufferDataForwardClustered::clear() {
 }
 
-void RenderForwardClustered::RenderBufferDataForwardClustered::configure(RID p_color_buffer, RID p_depth_buffer, int p_width, int p_height, RS::ViewportMSAA p_msaa) {
+void RenderForwardClustered::RenderBufferDataForwardClustered::configure(RID p_color_buffer, RID p_depth_buffer, int p_width, int p_height, RS::ViewportMSAA p_msaa, uint32_t p_views) {
```
for API consistency I would most likely rename p_views to p_view_count

BastiaanOlij replied:
👍

BastiaanOlij replied:
Done

reduz commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:
```diff
@@ -200,20 +201,25 @@ void RenderForwardClustered::RenderBufferDataForwardClustered::configure(RID p_c
 		tf.width = p_width;
 		tf.height = p_height;
-		tf.texture_type = RD::TEXTURE_TYPE_2D;
+		tf.array_layers = views;
```
likewise the same here layers to layer_count

BastiaanOlij replied:
Hmm, in this case I disagree.. it is view count here too, the problem is that array layers is used as the target for the different views. Else we'll just have the view_count turn into layer_count where it is actually reflecting view count.

BastiaanOlij replied:
I've added a remark that we need a layer for every view wherever we do this assignment.

BastiaanOlij commented:
Just a recap on my discussion with @reduz last night on Godot chat about the next steps. 

First in relation to our camera data. The plan we discussed yesterday is to change the data structure to the following:
- A single camera transform which in monoscopic is our camera transform and in stereo is a center point between the two eye transforms possibly moved back to support our clipping frustum.
- A single camera projection matrix which in monoscopic is our camera projection matrix and in stereo is a combined frustum that encompasses the two eyes as best as possible
- For stereo only, a projection matrix for each eye that combines the offset between the center position and the eye position with the eyes original projection matrix. 

These will be determined either at the start of `render_scene` in our culling class or in the function that currently handles the cameras and calls that `render_scene` function.

This approach means that we can keep a lot of the code that doesn't need to change for stereoscopic the same and more efficiently handle things that would otherwise need to be handled for stereoscopic just to accommodate the two eye positions. 

The logic that combines the two frustums for clipping will initially be the implementation we currently have in Godot 3 which only works properly if the two eyes share the same near and far clipping planes. But we do need to improve on this as not all devices have that guaranteed.

BastiaanOlij commented:
Looks like on Android CI we have to fix that we actually include Vulkan 1.1 headers.

![image](https://user-images.githubusercontent.com/1945449/118807638-e85a7600-b8eb-11eb-94dd-a637a553d0da.png)
Note the color change of the first box in this screenshot is to test accessing VIEW_INDEX in the shader.

All the basics are now working. Looking into improving the camera frustum merge logic and splitting all the logic into a few PRs so this thing doesn't grow to immense size. 

Also looking into removing the stereoscopic limit and having the XR Interface return any number of cameras so we can support quad displays and things like [LookingGlass](https://lookingglassfactory.com/)

Next step is cleaning up our XRInterface and seeing if we can get the OpenVR plugin to work.

BastiaanOlij commented:
Working on new frustum combine logic so this is currently broken

BastiaanOlij commented:
Finally fixed shadows and cleaned up the code creating the multiview variants of shaders, we now create them for all versions. 

![image](https://user-images.githubusercontent.com/1945449/119499600-3834b380-bdaa-11eb-953b-406dd0a8f79a.png)

BastiaanOlij commented:
Some more decisions in getting this read to merge our first version:
- we introduced a project setting whether VR support is enabled, this is off by default and prevents the multiview shaders from being compiled.
- we're going to remove support for the forward clustered renderer and put this into a separate PR and focus on getting VR to work in the mobile renderer first.

BastiaanOlij commented:
Added breaks compatibility as I'm changing the XRInterface to accommodate Vulkan as well as our related GDNative structs.

BastiaanOlij commented:
Rebased and fixed up the `Transform` -> `Transform3D` rename. 

I'm setting this ready for review. There is more to do around the XRInterface (for which I've started a new PR) and solving a lighting issue if the stereo camera isn't in the same plane (a future use case) but I think this is getting to a point where the work done so far needs to be evaluated and may even be ready to merge.

reduz commented on servers/rendering/renderer_rd/effects_rd.cpp:
```diff
@@ -1365,15 +1370,19 @@ void EffectsRD::cubemap_filter(RID p_source_cubemap, Vector<RID> p_dest_cubemap,
 }
 
-void EffectsRD::render_sky(RD::DrawListID p_list, float p_time, RID p_fb, RID p_samplers, RID p_fog, PipelineCacheRD *p_pipeline, RID p_uniform_set, RID p_texture_set, const CameraMatrix &p_camera, const Basis &p_orientation, float p_multiplier, const Vector3 &p_position) {
+void EffectsRD::render_sky(RD::DrawListID p_list, float p_time, RID p_fb, RID p_samplers, RID p_fog, PipelineCacheRD *p_pipeline, RID p_uniform_set, RID p_texture_set, const CameraMatrix &p_left_eye_projection, const CameraMatrix &p_right_eye_projection, const Basis &p_orientation, float p_multiplier, const Vector3 &p_position) {
```
For sky rendering, I think using left and right makes not much sense since for the most part you already made the most code N-views capable. I would fix this for consistency.

reduz commented on servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp:
```diff
@@ -656,6 +657,11 @@ void SceneShaderForwardClustered::init(RendererStorageRD *p_storage, const Strin
 
+		// not implemented but need these just in case code is in the shaders
+		actions.renames["VIEW_INDEX"] = "0";
+		actions.renames["VIEW_MONO_LEFT"] = "0";
```
I would kinda expect that VIEW_INDEX 0 and 1 should be enough. Also should it not be MULTIVIEW_INDEX ? I am not sure this is very useful outside multiview.

BastiaanOlij replied:
`VIEW_INDEX` is the actual variable, it's only 0 here because its turned off in the forward clustered, this code is only there so the compiler doesn't throw up. See the mobile implementation for the proper code.

reduz commented on servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:
```diff
@@ -488,14 +494,18 @@ void RenderForwardMobile::_render_scene(RenderDataRD *p_render_data, const Color
 		RENDER_TIMESTAMP("Render Sky");
 
-		CameraMatrix projection = p_render_data->cam_projection;
+		// TODO change this to supporting more then 2 view projections
```
You already made most code N-views already, so I would go the extra mile and convert this one too.

reduz commented on servers/rendering/renderer_rd/forward_mobile/scene_shader_forward_mobile.cpp:
```diff
@@ -585,10 +544,23 @@ void SceneShaderForwardMobile::init(RendererStorageRD *p_storage, const String p
+
 		shader.initialize(shader_versions, p_defines);
+
+		bool xr_enabled = GLOBAL_GET("rendering/xr/enabled");
```
I would most likely have this bool in a bit more central place, like in the compositor, where you can query whether multiview is enabled.

BastiaanOlij replied:
You mean like read it once and keep it as a member variable of the compositor that we can read from the other places?

BastiaanOlij replied:
Ok, I ran into a problem with the compositor because one of the places this setting is accessed is in the constructor of `EffectsRD`. This object is constructed when storage is constructed, which in turn is constructed within the compositors constructor, which means we don't have access to our variable yet unless we want to start passing pointers to the constructors. 

@reduz I think I'll keep this one for later once we discuss an alternative

BastiaanOlij replied:
Ok, now accessing through the singleton on `RendererCompositorRD`, that did the trick :)

reduz commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -2186,142 +2186,89 @@ bool RendererSceneCull::_light_instance_update_shadow(Instance *p_instance, cons
-		} break;
-	}
+	// Setup Camera(s)
+	if (p_interface.is_valid()) {
```
Nitpick here, I would use p_interface.is_null() and swap the ifs() so the most commonly used case goes first, to make the code more readable. Also please put a comment specifying this is mostly for XR, so the code is more readable.

reduz commented on servers/rendering/renderer_scene_cull.h:
```diff
@@ -952,11 +952,10 @@ class RendererSceneCull : public RendererScene {
 
 	bool _render_reflection_probe_step(Instance *p_instance, int p_step);
-	void _render_scene(const Transform3D &p_cam_transform, const CameraMatrix &p_cam_projection, bool p_cam_orthogonal, bool p_cam_vaspect, RID p_render_buffers, RID p_environment, RID p_force_camera_effects, uint32_t p_visible_layers, RID p_scenario, RID p_viewport, RID p_shadow_atlas, RID p_reflection_probe, int p_reflection_probe_pass, float p_screen_lod_threshold, bool p_using_shadows = true);
+	void _render_scene(const RendererSceneRender::ViewData *p_view_data, RID p_render_buffers, RID p_environment, RID p_force_camera_effects, uint32_t p_visible_layers, RID p_scenario, RID p_viewport, RID p_shadow_atlas, RID p_reflection_probe, int p_reflection_probe_pass, float p_screen_lod_threshold, bool p_using_shadows = true);
```
I would probably rename ViewData to CameraData, since you have view independent information there, that will make the code more readable.

BastiaanOlij replied:
I named it viewdata with the idea that we may some day have more then 2 cameras and it directs what is shown on the views, but I'm happy to rename it

reduz commented on servers/rendering/renderer_scene_render.cpp:
```diff
@@ -29,3 +29,153 @@
 
 #include "renderer_scene_render.h"
+
+void RendererSceneRender::ViewData::set_single_view(const Transform3D p_transform, const CameraMatrix p_projection, bool p_is_ortogonal, bool p_vaspect) {
```
I would probably also rename those into set_camera and set_multiview_camera, to make sure multiview is the exception.

BastiaanOlij replied:
Sounds good to me, I'll make it so

BastiaanOlij commented on servers/rendering/renderer_scene_render.h:
```diff
@@ -35,6 +35,8 @@
 #include "core/templates/paged_array.h"
 #include "servers/rendering/renderer_storage.h"
 
+#define RENDER_MAX_VIEWS 2
```
After discussing with @reduz , see if we can add this to the enums, maybe part of `RendererCompositor` so it's cleaner. I had problems with this before but may have been due to unrelated issues.

reduz commented:
Should be ready to merge once it passes CI

reduz commented:
The only problem I see to this is that its still missing the usage of the proper eye vector on each view, so reflections and specular blobs will most likely not be correct. We can add this later as a separate PR if you wish.

akien-mga commented:
> The only problem I see to this is that its still missing the usage of the proper eye vector on each view, so reflections and specular blobs will most likely not be correct. We can add this later as a separate PR if you wish.

Yeah let's do this as a separate PR I think, so that we can merge this and @JFonS' #48847.

akien-mga commented:
Thanks!
